src/AppInstallerRepositoryCore/Rest/Schema/1_0/Interface.cpp

```diff
@@ -89,115 +78,115 @@ namespace AppInstaller::Repository::Rest::Schema::V1_0
-                std::optional<std::string> channel = GetStringFromJsonValue(versionItem.at(GetJsonKeyNameString(Channel)));
+                std::vector<AppInstaller::Manifest::ValidationError> validationErrors =
+                    AppInstaller::Manifest::ValidateManifest(manifest.value());
 
-                if (!version.has_value())
+                int errors = 0;
+                for (auto& error : validationErrors)
                 {
-                    AICLI_LOG(Repo, Verbose, << "Received incomplete package version. Skipping version from package: " << packageId.value());
-                    continue;
+                    if (error.ErrorLevel == Manifest::ValidationError::Level::Error)
+                    {
+                        AICLI_LOG(Repo, Verbose, << "Received manifest contains validation error: " << error.Message);
+                        errors++;
+                    }
                 }
 
-                versionList.emplace_back(AppInstaller::Utility::VersionAndChannel{ std::move(version.value()), std::move(channel.value_or("")) });
+                if (errors > 0)
+                {
+                    AICLI_LOG(Repo, Verbose, << "Received invalid manifest. Skipping");
+                    return {};
+                }
             }
+        }
 
-            if (versionList.size() == 0)
-            {
-                AICLI_LOG(Repo, Verbose, << "Received no valid versions. Skipping package: " << packageId.value());
-                continue;
-            }
+        return manifest;
+    }
 
-            PackageInfo packageInfo = PackageInfo{ std::move(packageId.value()), std::move(packageName.value()), std::move(publisher.value()) };
-            Package package = Package{ std::move(packageInfo), std::move(versionList) };
-            result.Matches.emplace_back(std::move(package));
+    bool Interface::MeetsOptimizedSearchCriteria(const SearchRequest& request) const
+    {
+        if (!request.Query.has_value() && request.Inclusions.size() == 0 &&
+            request.Filters.size() == 1 && request.Filters[0].Field == PackageMatchField::Id &&
+            request.Filters[0].Type == MatchType::Exact)
+        {
+            AICLI_LOG(Repo, Verbose, << "Search request meets optimized search criteria.");
+            return true;
         }
 
-        return result;
+        return false;
     }
 
-    std::optional<Manifest::Manifest> Interface::GetManifestByVersion(const std::string& packageId, const std::string& version, const std::string& channel) const
+    IRestClient::SearchResult Interface::OptimizedSearch(const SearchRequest& request) const
     {
-        HttpClientHelper clientHelper{ GetManifestByVersionEndpoint(m_restApiUri, packageId, version, channel) };
-        web::json::value jsonObject = clientHelper.HandleGet();
+        SearchResult searchResult;
+        std::optional<Manifest::Manifest> manifestResult = GetManifestByVersion(request.Filters[0].Value, {}, {});

```

Comment:
I thought that this API would return *ALL* manifests if no version was supplied.

Reply:
you are correct. I will add a method that gets all the returned versions from the endpoint and will have methods call that and decide what to do with the multiple versions case by case.

---
In reply to: [596382444](https://github.com/microsoft/winget-cli/pull/800#discussion_r596382444) [](ancestors = 596382444)